### PR TITLE
add json2wal & fix wal2json

### PR DESCRIPTION
**Pull request by bradyjoestar**
I am try to solve the issue #1954 

In wal2json file , it use json.Marshal to convert msg to json.

If We want to unMarshal json to TimedWALMessage,we'd better use go-amino to convert msg to json file.

* [Y] Updated all relevant documentation in docs
* [Y ] Updated all code comments where relevant
* [] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: scripts/json2wal/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+func init() {
+	cs.RegisterConsensusMessages(cdc)
+	cs.RegisterWALMessages(cdc)
+	types.RegisterBlockAmino(cdc)
+}
+
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Println("missing arguments: Usage:json2wal <path-to-JSON>  <path-to-wal>")
```
Comment (melekes) [inline scripts/json2wal/main.go]: `fmt.Fprintln(os.Stderr, "missing`

File: scripts/json2wal/main.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+	for {
+		msgJson, _, err := br.ReadLine()
+		if err == io.EOF{
+			break
+		}else if err != nil {
+			panic(fmt.Errorf("failed to read file: %v", err))
+		}
+		// ignore the ENDHEIGHT in json.File
+		if strings.HasPrefix(string(msgJson),"ENDHEIGHT"){
```
Comment (melekes) [inline scripts/json2wal/main.go]: why do we ignore it? it should be present in the output file, no?

File: scripts/json2wal/main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		os.Exit(1)
+	}
+
+	f, err := os.Open(os.Args[1])
+	if err != nil {
+		panic(fmt.Errorf("failed to open WAL file: %v", err))
+	}
+	defer f.Close()
+
+	walFile, err := os.OpenFile(os.Args[2],os.O_APPEND|os.O_WRONLY|os.O_CREATE,0666)
```
Comment (melekes) [inline scripts/json2wal/main.go]: don't think we should append. better crash if file already exists - `O_EXCL`

File: scripts/json2wal/main.go (hunk 1 above)
Comment (bradyjoestar) [reply scripts/json2wal/main.go]: ok,I will update.

Comment (bradyjoestar) [reply scripts/json2wal/main.go]: ok.

File: scripts/json2wal/main.go (hunk 3 above)
Comment (bradyjoestar) [reply scripts/json2wal/main.go]: ok.

File: scripts/json2wal/main.go (hunk 2 above)
Comment (bradyjoestar) [reply scripts/json2wal/main.go]: I found the following code in the wal2json.file
```
if err == nil {
			if end, ok := msg.Msg.(cs.EndHeightMessage); ok {
				_, err = os.Stdout.Write([]byte(fmt.Sprintf("ENDHEIGHT %d\n", end.Height))) // nolint: errcheck, gas
			}
		}
```
so I ignore the "ENDHEIGHT"

File: scripts/json2wal/main.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func main() {
 	}
 	defer f.Close()
 
-	walFile, err := os.OpenFile(os.Args[2],os.O_APPEND|os.O_WRONLY|os.O_CREATE,0666)
+	walFile, err := os.OpenFile(os.Args[2],os.O_EXCL|os.O_WRONLY,0666)
```
Comment (melekes) [inline scripts/json2wal/main.go]: why did you remove `O_CREATE`?

File: scripts/json2wal/main.go (hunk 2 above)
Comment (melekes) [reply scripts/json2wal/main.go]: Have you tried converting existing wal to json, converting json to wal and running it? Does it work?

It should be more like:

```go
if strings.HasPrefix(string(msgJson),"ENDHEIGHT"){
h := extractHeight(string(msgJson))
err = dec.Encode(&cs.EndHeightMessage{h})
		if err != nil{
			panic(fmt.Errorf("failed to encode msg: %v", err))
		}
}
```

File: scripts/json2wal/main.go (hunk 4 above)
Comment (bradyjoestar) [reply scripts/json2wal/main.go]: oh,my careless

File: scripts/json2wal/main.go (hunk 2 above)
Comment (bradyjoestar) [reply scripts/json2wal/main.go]: Yeah, I use 'md5sum' command to check the hash of existing wal and generated wal on ubuntu,and it matched.I will upload my wal right now.

Comment (bradyjoestar) [reply scripts/json2wal/main.go]: If there are EndHeightMessage in json file， go-amino will help us  write it into generated wal.
I just ignore the prefix of 'ENDHEIGHT',not 
```
{"time":"2018-08-08T11:28:00.372359633Z","msg":{"type":"tendermint/wal/EndHeightMessage","value":{"height":"2"}}}

```

Comment (bradyjoestar) [reply scripts/json2wal/main.go]: generated json file slice
```
{"time":"2018-08-08T11:28:00.285064707Z","msg":{"type":"tendermint/wal/EventDataRoundState","value":{"height":"2","round":"0","step":"RoundStepCommit"}}}
{"time":"2018-08-08T11:28:00.372359633Z","msg":{"type":"tendermint/wal/EndHeightMessage","value":{"height":"2"}}}
ENDHEIGHT 2
```

Comment (melekes) [reply scripts/json2wal/main.go]: Oh, right. Awesome. Thank you

**Comment by melekes** [comment]
I think you've accidentally checked-in scripts/wal2JsonTest binary

**Comment by ValarDragon** [comment]
Want to add an entry to the changelog? (Just to let ppl know its back)

**Comment by melekes** [comment]
I'll add a changelog entry myself.
